`Clear-Site-Data: "cache"` should only suppress the next navigation snapshot in case of same-origin navigation

cdumez opened the pull request:
#### ff41eac48894511fe1be7de026034a39a27e2f2b
<pre>
`Clear-Site-Data: &quot;cache&quot;` should only suppress the next navigation snapshot in case of same-origin navigation
<a href="https://bugs.webkit.org/show_bug.cgi?id=251855">https://bugs.webkit.org/show_bug.cgi?id=251855</a>

Reviewed by Sihui Liu.

`Clear-Site-Data: &quot;cache&quot;` should only suppress the next navigation snapshot in
case of same-origin navigation.

So if you navigate from A1 to B, and then to A2 (which serves the
Clear-Site-Data HTTP header). You&apos;d expect the snapshot for A1 to get deleted
but not the one for B. This is because A2 served the header and we should thus
only delete snapshots for origin A.

Previously, we would prevent the creation of the snapshot for B when navigating
to A2. This patch fixes this.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::deleteWebsiteDataInWebProcessesForOrigin):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SnapshotStore.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/260034@main">https://commits.webkit.org/260034@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a2be6230d4ec090c9214eba21841b21356b0ae1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106807 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115990 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110710 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/17323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7033 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98995 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112575 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/17323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96112 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/17323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/27764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/9007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9574 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48663 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6925 "Found 1 new test failure: fast/text/web-font-load-invisible-during-loading.html (failure)") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11100 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3741 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (approved):
r=me
It might be better to state in the commit message that B's snapshot is actually not created (instead of being deleted)

cdumez committed:
`Clear-Site-Data: "cache"` should only suppress the next navigation snapshot in case of same-origin navigation
https://bugs.webkit.org/show_bug.cgi?id=251855

Reviewed by Sihui Liu.

`Clear-Site-Data: "cache"` should only suppress the next navigation snapshot in
case of same-origin navigation.

So if you navigate from A1 to B, and then to A2 (which serves the
Clear-Site-Data HTTP header). You'd expect the snapshot for A1 to get deleted
but not the one for B. This is because A2 served the header and we should thus
only delete snapshots for origin A.

Previously, we would prevent the creation of the snapshot for B when navigating
to A2. This patch fixes this.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::deleteWebsiteDataInWebProcessesForOrigin):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SnapshotStore.mm:
(TEST):

Canonical link: https://commits.webkit.org/260034@main
